pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseSingleStageBrokerRequestHandler.java

```diff
@@ -755,19 +755,14 @@ protected BrokerResponse doHandleRequest(long requestId, String query, SqlNodeAn
     if (isQueryCancellationEnabled()) {
       // Start to track the running query for cancellation just before sending it out to servers to avoid any
       // potential failures that could happen before sending it out, like failures to calculate the routing table etc.
-      // TODO: Even tracking the query as late as here, a potential race condition between calling cancel API and
-      //       query being sent out to servers can still happen. If cancel request arrives earlier than query being
-      //       sent out to servers, the servers miss the cancel request and continue to run the queries. The users
-      //       can always list the running queries and cancel query again until it ends. Just that such race
-      //       condition makes cancel API less reliable. This should be rare as it assumes sending queries out to
-      //       servers takes time, but will address later if needed.
       String clientRequestId = extractClientRequestId(sqlNodeAndOptions);
-      onQueryStart(
-          requestId, clientRequestId, query, new QueryServers(query, offlineRoutingTable, realtimeRoutingTable));
+      final Map<ServerInstance, ServerRouteInfo> finalOfflineRoutingTable = offlineRoutingTable;
+      final Map<ServerInstance, ServerRouteInfo> finalRealtimeRoutingTable = realtimeRoutingTable;
```

Comment:
nit: we don't usually add final to variables. There is no advantage on doing so and makes lines longer. Sometimes it may be useful to make it clear that it won't be changed, but that should be the general case so unless you think it is especially interesting to show that (ie a final variable declared on a method where tons of other variables are modified) I think it is better to not add the final qualifier